src/ingress/internal.rs

```diff
@@ -0,0 +1,168 @@
+use std::time::Duration;
+
+use reqwest::{header::HeaderMap, Url};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+use super::{
+    request::{IngressRequestOptions, SendResponse, SendStatus},
+    result::{IngressResultOptions, ResultOp, ResultTarget},
+};
+use crate::{context::RequestTarget, errors::TerminalError};
+
+const IDEMPOTENCY_KEY_HEADER: &str = "Idempotency-Key";
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct SendResponseSchema {
+    invocation_id: String,
+    status: SendStatusSchema,
+}
+
+#[derive(Deserialize)]
+enum SendStatusSchema {
+    Accepted,
+    PreviouslyAccepted,
+}
+
+impl From<SendStatusSchema> for SendStatus {
+    fn from(value: SendStatusSchema) -> Self {
+        match value {
+            SendStatusSchema::Accepted => SendStatus::Accepted,
+            SendStatusSchema::PreviouslyAccepted => SendStatus::PreviouslyAccepted,
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct TerminalErrorSchema {
+    code: Option<u16>,
+    message: String,
+}
+
+pub(super) struct IngressInternal {
+    pub(super) client: reqwest::Client,
+    pub(super) url: Url,
+    pub(super) headers: HeaderMap,
+}
+
+impl IngressInternal {
+    pub(super) async fn call<Req: Serialize, Res: DeserializeOwned>(
+        &self,
+        target: RequestTarget,
+        req: Req,
+        opts: IngressRequestOptions,
+    ) -> Result<Result<Res, TerminalError>, reqwest::Error> {
```

Comment:
Rather than using `Result<Result<Res, TerminalError>, reqwest::Error>` around, I would define a new error enum type `IngressClientError` with two variants: `Terminal(TerminalError)` or `Http(reqwest::Error)`.